wasm/caller.go

```diff
@@ -0,0 +1,59 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build wasm
+
+package wasm
+
+import (
+	"unsafe"
+)
+
+// HostFunc is the function type for the imported functions from the host.
+//
+// The arguments are:
+// (1) a pointer to the address where the command should be written
+// (2) the size of allocated memory.
+//
+// The return value indicates the size of the allocated response in bytes. If the
+// response is larger than the allocated memory, the caller should reallocate the
+// memory and call the function again.
+type HostFunc func(ptr unsafe.Pointer, size uint32) uint32
+
+type HostCaller struct {
+	Func HostFunc
+}
+
+// Call calls the function from the host 2 times max, is the buffer size is not
+// enough the first time its called, it will be resized the second call.
+// returns the buffer, the command size, and the error.
+func (fc *HostCaller) Call(buf []byte, size int) ([]byte, uint32, error) {
```

Comment:
3 comments here:
- I don't see a reason for this to be a type method, maybe change this into a static method instead?
- I wouldn't necessarily expose it outside of this package.
- Do we need an explicit `size` parameter? Can we instead just use `len(buf)`?

Something like this maybe:
```suggestion
func hostCall(fn hostFunc, buf []byte) ([]byte, int, error) {
```